src/qibo/models.py

```diff
@@ -138,6 +139,29 @@ def _DistributedQFT(nqubits: int,
     return circuit
 
 
+def parallel_loss(params, resource):
+    return resource().loss(params)
+
+
+class vqe_optimparallel_resources(optimparallel_resources):
+
+    hamiltonian = None
+    circuits = None
+
+    def setup(self, circuit, hamiltonian, parallel):
+        self.hamiltonian = hamiltonian
+        max_workers = self.get_max_workers(parallel)
+        self.circuits = [circuit.copy(deep=True) for _ in range(max_workers)]
+
+    def loss(self, params, *args):
+        import multiprocessing as mp
+        thread_id = int(mp.current_process().name[-1])-1
+        circuit = self.circuits[thread_id]
```

Comment:
We could also use a dictionary for `self.circuits` to avoid this `int` conversion of `thread_id`, however I don't think it is much of a difference.